pkg/urbit/worker/mars.c

```diff
@@ -582,30 +584,30 @@ _mars_flush(u3_mars* mar_u)
 {
 top:
   {
-    u3_gift* gif_u = mar_u->gif_u.ext_u;
+    u3_gift* gif_u = c3_lode_data(c3_list_peekf(mar_u->gif_u));
 
     //  XX gather i/o
     //
     while (  gif_u
           && (  (u3_gift_rest_e == gif_u->sat_e)
-             || (gif_u->eve_d <= mar_u->log_u->dun_d)) )
+             || (gif_u->eve_d <= u3_saga_last_commit(mar_u->log_u)) ) )
     {
       u3_newt_send(mar_u->out_u, gif_u->len_d, gif_u->hun_y);
 
-      mar_u->gif_u.ext_u = gif_u->nex_u;
-      c3_free(gif_u);
-      gif_u = mar_u->gif_u.ext_u;
-    }
-
-    if ( !mar_u->gif_u.ext_u ) {
-      mar_u->gif_u.ent_u = 0;
+      c3_free(c3_list_popf(mar_u->gif_u));
+      gif_u = c3_lode_data(c3_list_peekf(mar_u->gif_u));
     }
   }
 
   if (  (u3_mars_work_e != mar_u->sat_e)
-     && (mar_u->log_u->dun_d == mar_u->dun_d) )
+     && (u3_saga_last_commit(mar_u->log_u) == mar_u->dun_d) )
   {
     if ( u3_mars_save_e == mar_u->sat_e ) {
+      if ( u3_saga_needs_rollover(mar_u->log_u)
+          && !u3_saga_rollover(mar_u->log_u) )
+      {
+        fprintf(stderr, "mars: failed to rollover to new epoch\r\n");
+      }
```

Comment:
Again, this movement of event log specific logic out of `saga` and into `mars` worries me. If we're not careful, instead of two modules with clear responsibilities, we'll end up with a single, much bigger module that's much more difficult to understand. `mars` is the state machine, and while the event log is a vital component of that state machine, it's not small enough to be subsumed within the state machine, especially if we hope to be able to swap out event log implementations for the sake of hosting.